### DISCUSSION TITLE
The problem of parallel computing to improve efficiency

### DISCUSSION THREAD

**Discussion by buzhiqiuyuebai** [SeisSol/SeisSol#704] [Q&A] [answered]
I run a SeisSol example WP2_LOH3 For parallelism testing，but when I improve the mpirun cores when running the example, I find there is no obvious increase in the parallel efficiency of the example, 
![image](https://user-images.githubusercontent.com/114474894/197016101-8e9a5e90-ac7d-4b5e-b345-f62d555f2ee1.png)
and the following picture is I get when I test the mpirun cores and the time used in this example,  I have modified the size of the volume and run time is 0.02s, I think it should have  an obvious improvement in calculate time, Do you know why this happen, Thanks  in advance.
![image](https://user-images.githubusercontent.com/114474894/197017984-cb32b4ec-f143-4dd0-bdf5-6a54e0e4c180.png)

**Comment by Thomas-Ulrich** [comment, marked as answer]
Hi,
On which computer are you running? 
How many nodes and cores does have the system? 

Le 20 octobre 2022 19:30:47 GMT+02:00, buzhiqiuyuebai ***@***.***> a écrit :

**Reply by buzhiqiuyuebai** [reply]
> Hi, On which computer are you running? How many nodes and cores does have the system? Le 20 octobre 2022 19:30:47 GMT+02:00, buzhiqiuyuebai ***@***.***> a écrit :
> […](#)
> I run a SeisSol example WP2_LOH3 For parallelism testing，but when I improve the mpirun cores when running the example, I find there is no obvious increase in the parallel efficiency of the example, > ![image](https://user-images.githubusercontent.com/114474894/197016101-8e9a5e90-ac7d-4b5e-b345-f62d555f2ee1.png) and the following picture is I get when I test the mpirun cores and the time used in this example, I have modified the size of the volume and run time is 0.02s, I think it should have an obvious improvement in calculate time, Do you know why this happen, Thanks in advance. ![image](https://user-images.githubusercontent.com/114474894/197017984-cb32b4ec-f143-4dd0-bdf5-6a54e0e4c180.png) -- > Reply to this email directly or view it on GitHub: #704 You are receiving this because you are subscribed to this thread. Message ID: ***@***.***>

This is the CPU parameters
![image](https://user-images.githubusercontent.com/114474894/197147334-f6f1b93a-6f64-469b-b13b-a3c3aac7a034.png)
I have 4 nodes and 112 cores in my computers.

**Reply by Thomas-Ulrich** [reply]
ok then you can do things like:
```
export OMP_NUM_THREADS=28 
mpirun -n 4 seissol_exe parameters.par
```
as 28*4=112.
(or let say  ```export OMP_NUM_THREADS=56``` with hyperthreading).
But you should not use more $OMP_NUM_THREADS * ranks than twice the number of CPUs.

**Reply by buzhiqiuyuebai** [reply]
> ok then you can do things like:
> 
> ```
> export OMP_NUM_THREADS=28 
> mpirun -n 4 seissol_exe parameters.par
> ```
> 
> as 28*4=112. (or let say `export OMP_NUM_THREADS=56` with hyperthreading). But you should not use more $OMP_NUM_THREADS * ranks than twice the number of CPUs.

Thank you for your help, why set `export OMP_NUM_THREADS=28` ,What's the benefit of doing this，and I find when the number of `OMP_NUM_THREADS` is same, increase the number of MPI ranks, it  seems no obvious increase in the parallel efficiency, as the picture belong:
![image](https://user-images.githubusercontent.com/114474894/197388084-163922f0-cd2b-4bce-b0b7-95fc66fac902.png)

when the number of `MPI ranks` is same, increase the number of `OMP_NUM_THREADS`，When using 16 MPI ranks, increasing the number of `OMP_NUM_THREADS` can greatly improve the efficiency, but when using 8 and 4 MPI ranks, increasing the number of `OMP_NUM_THREADS` reduces the efficiency.
![image](https://user-images.githubusercontent.com/114474894/197388102-237f7793-5516-44cf-a894-1191868f06aa.png)

and when using $OMP_NUM_THREADS * ranks=56*2, I encounter this problem.
![image](https://user-images.githubusercontent.com/114474894/197386264-76bc1000-07cc-49bb-94d9-cfdbd3cb40f1.png)

**Reply by smallriver666** [reply]
> ok then you can do things like:
> 
> ```
> export OMP_NUM_THREADS=28 
> mpirun -n 4 seissol_exe parameters.par
> ```
> 
> as 28*4=112. (or let say `export OMP_NUM_THREADS=56` with hyperthreading). But you should not use more $OMP_NUM_THREADS * ranks than twice the number of CPUs.

According to your experience, how to allocate the relationship between MPI and OpenMP is the most efficient? Should the number of OpenMP be set larger and the number of mpi processes be less, or should the number of OpenMP be set smaller and the number of mpi processes be more?
